### PR TITLE
Remove ecdsa dependency, use cryptography only II

### PR DESCRIPTION
**Pull request by onvej-sl**
This replaces https://github.com/trezor/python-slip10/pull/2. Unlike that, I've tried to use the cryptography package as much as possible.

### CODE REVIEW

File: slip10/utils.py [hunk 1]
```diff
@@ -104,31 +104,64 @@ def derive_public_child(self, pubkey, chaincode, index):
...
+
+        if x1 == x2 and y1 != y2:
+            return None
+
+        if self.curve.name == "secp256k1":
+            p = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F
+        elif self.curve.name == "secp256r1":
+            p = 0xFFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF
+        else:
+            raise NotImplementedError(f"Curve {self.curve.name} is not supported")
```
Comment (onvej-sl) [inline slip10/utils.py]: Since the implementation of elliptic curves in the cryptography package is only a wrapper over openssl, the package doesn't even export the curves' moduli.

File: pyproject.toml [hunk 2]
```diff
@@ -13,9 +13,8 @@ repository = "https://github.com/trezor/python-slip10"
 keywords = ["bitcoin", "slip10", "hdwallet"]
 
 [tool.poetry.dependencies]
-cryptography = "*"
-ecdsa = "*"
-python = ">=3.8,<4.0"
+cryptography = ">=45"
+python = ">3.9.0,<3.9.1 || >3.9.1,<3.14.0"
```
Comment (andrewkozlik) [inline pyproject.toml]: 1. ISTM that `>3.9.0,<3.9.1` is not satisfiable, so only the latter condition makes sense. Or am I misunderstanding the syntax?
2. Curious, why is 3.14 excluded?
3. Let's update `.github/workflows/python-package.yml` accordingly.

Comment (onvej-sl) [reply pyproject.toml]: > 1. ISTM that >3.9.0,<3.9.1 is not satisfiable, so only the latter condition makes sense. Or am I misunderstanding the syntax?
> 2. Curious, why is 3.14 excluded?

I don't consider this to be the final solution. I only did what poetry advised me to do in order to upgrade the cryptography package.

File: tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import pytest
+
+from slip10.utils import SECP256K1, SECP256R1
+
+
+def test_secp256k1():
+    generator = SECP256K1.privkey_to_pubkey((1).to_bytes(32, "big"))
```
Comment (andrewkozlik) [inline tests/test_utils.py]: It should be possible to remove the code duplication by using something like this:
```suggestion
@pytest.mark.parametrize("curve", (SECP256K1, SECP256R1))
def test_secp_curves(curve):
    generator = curve.privkey_to_pubkey((1).to_bytes(32, "big"))
    // ... replace SECP256K1 with curve everywhere
```

Comment (onvej-sl) [reply tests/test_utils.py]: Done in https://github.com/trezor/python-slip10/pull/4/commits/3ab780e3bba7a22f0355d2f7869c7c6a59959fbe.

File: slip10/utils.py [hunk 4]
```diff
@@ -104,31 +106,58 @@ def derive_public_child(self, pubkey, chaincode, index):
...
             ).digest()
-        return point.to_bytes("compressed"), payload[32:]
+        return point, payload[32:]
 
     def privkey_is_valid(self, privkey):
         key = int.from_bytes(privkey, "big")
-        return 0 < key < self.curve.order
+        return 0 < key < self.order
+
+    def add_points(self, first: bytes, second: bytes) -> bytes:
```
Comment (andrewkozlik) [inline slip10/utils.py:124]: How does this function handle if one of the inputs is a point at infinity? Does `from_encoded_point()` raise an exception? I think the only way to get infinity as an input is if the tweak `payload[:32]` comes out zero, which is practically impossible to observe. Maybe `privkey_to_pubkey()` already raises an exception in that case. We should at least have a comment explaining how this function behaves for infinity inputs and why it's OK.

Comment (onvej-sl) [reply slip10/utils.py:124]: Besides that, I also realized that `privkey_to_pubkey()` raises an exception if `privkey` is greater than the curve's order, which would break [this test vector](https://github.com/satoshilabs/slips/blob/master/slip-0010.md#test-derivation-retry-for-nist256p1), if we had it included in our tests.

Comment (andrewkozlik) [reply slip10/utils.py:124]: We have that test vector included in our tests, <s>but it tests retry in [master key generation](https://github.com/satoshilabs/slips/blob/master/slip-0010.md#master-key-generation). We should generate a test vector for retry in child key derivation too and add it to the spec.</s>

Comment (onvej-sl) [reply slip10/utils.py:124]: > Besides that, I also realized that privkey_to_pubkey() raises an exception if privkey is greater than the curve's order

I fixed that https://github.com/trezor/python-slip10/pull/4/commits/e19712778297252a2e7e4dd8868163719167ad55 by implementing `multiply_generator()` that accepts all integer scalars.

> How does this function handle if one of the inputs is a point at infinity?

I fixed that in https://github.com/trezor/python-slip10/pull/4/commits/e19712778297252a2e7e4dd8868163719167ad55 by handling these edge case in both `multiply_generator()` and `add_points()`. Although the probability of this happening is infinitesimal, I think it's more readable to handle it explicitly.

> We have that test vector included in our tests

You are right. The problem is that the test doesn't test public key to public key derivation. I added another test in https://github.com/trezor/python-slip10/pull/4/changes/66d62db44f1b845a31332a5042934f3627ccff62.

File: tests/test_slip10.py [hunk 5]
```diff
@@ -914,3 +913,22 @@ def test_slip10_vectors():
...
+        chaincode, pubkey=pubkey, curve_name="secp256r1"
+    ).get_extended_pubkey_from_path("m/33941")
+    assert (
+        chaincode.hex()
+        == "9e87fe95031f14736774cd82f25fd885065cb7c358c1edf813c72af535e83071"
+    )
+    assert (
+        pubkey.hex()
+        == "0235bfee614c0d5b2cae260000bb1d0d84b270099ad790022c1ae0b2e782efe120"
+    )
```
Comment (andrewkozlik) [inline tests/test_slip10.py:934]: This works, although ideally we should have a test that takes the `path` in each test vector, converts it to a list `_deriv_path_str_to_list(path)` and counts how many indices at the end are non-hardened. Then it should derive the node up to the last hardened index using `get_child_from_path()` and from there derive using only public keys. Basically the above test but automated. (Doesn't need to be done in this PR.)

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    triple_generator = curve.multiply_generator(3)
+
+    assert curve.add_points(bytes.fromhex("00"), generator) == generator
+    assert curve.add_points(generator, bytes.fromhex("00")) == generator
+    assert curve.add_points(generator, minus_generator) == bytes.fromhex("00")
+    assert curve.add_points(generator, generator) == double_generator
+    assert curve.add_points(generator, double_generator) == triple_generator
+
+    assert curve.multiply_generator(0) == bytes.fromhex("00")
+    assert curve.multiply_generator(curve.order) == bytes.fromhex("00")
```
Comment (andrewkozlik) [inline tests/test_utils.py]: nit: How about using `curve.point_at_infinity` instead of `bytes.fromhex("00")`?

Comment (onvej-sl) [reply tests/test_utils.py]: Done in https://github.com/trezor/python-slip10/pull/4/commits/b649486079642bac3ae2f1f2b4e2c29fa24cb525.

**Review by andrewkozlik — APPROVED** [review]
